crypto/evp/keymgmt_lib.c

```diff
@@ -412,3 +412,51 @@ void *evp_keymgmt_util_gen(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
 
     return keydata;
 }
+
+/*
+ * Returns the same numbers as EVP_PKEY_get_default_digest_name()
+ * When the string from the EVP_KEYMGMT implementation is "", we use
+ * SN_undef, since that corresponds to what EVP_PKEY_get_default_nid()
+ * returns for no digest.
+ */
+int evp_keymgmt_util_get_deflt_digest_name(EVP_KEYMGMT *keymgmt,
+                                           void *keydata,
+                                           char *mdname, size_t mdname_sz)
+{
+    OSSL_PARAM params[3];
+    char mddefault[100] = "";
+    char mdmandatory[100] = "";
+    char *result = NULL;
+    int rv = -2;
+
+    params[0] =
+        OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_DEFAULT_DIGEST,
+                                         mddefault, sizeof(mddefault));
+    params[0].return_size = sizeof(mddefault) + 1;
```

Comment:
Setting the return size here will have no effect.  The return size is set by the recipient.

This isn't a great solution to the _did the other end set this_ question that we've thus far not been able to answered.

Reply:
> Setting the return size here will have no effect. The return size is set by the recipient.

... *if* the recipient supports this parameter, and in that case, it will be set to something other than this initial value.

Reply:
I noted this, I don't like this way of detecting a parameter being set.

It's error prone -- the return size is set but the buffer not if the return size is larger than the data length.  So a legitimate return of size `sizeof(mddefault) + 1` could be a problem -- it isn't here because of another check but I think it is a bad precedent to set.

I'd consider using `(size_t)-1` would be better.  Best would be for the helper functions to support what is required.  The construct calls clearing whatever sentinel is used, the set calls flag it and an additional function to check.

Reply:
> So a legitimate return of size `sizeof(mddefault) + 1` could be a problem

If `p.data_size` is `sizeof(mddefault)`, I don't see how that would be a legitimate return size...  that would mean that the receiving end has overwritten the buffer.

> I'd consider using `(size_t)-1` would be better.

I've considered that, but couldn't shake the feeling that there may be platforms where that's a perfectly legitimate size that doesn't necessarily eat all available memory (i.e. where `size_t` is smaller in size than `void *`).  It's *unlikely* that we'll hit that kind of size, at least for now, but...

> Best would be for the helper functions to support what is required.  The construct calls clearing whatever sentinel is used, the set calls flag it and an additional function to check.

I have nothing against that idea.

Reply:
> If p.data_size is sizeof(mddefault), I don't see how that would be a legitimate return size...

If the return size is larger than the buffer size, the return size is set but no data is returned.  The function also returns 0, so it's not an issue here.

I'll try make a stab at an API tomorrow.  No promises.

Reply:
Don't make this a blocker for the upcoming alpha.

Reply:
Interestingly, a compliant C99 implementation is allowed to set `SIZE_MAX` as small as 65535.

Reply:
I've used processors where even that would be way too large :)

Reply:
Kind of moot given the delay but refer #11588 for an alternative

Reply:
I've come to terms with the `(size_t)-1` value.  It will probably be eons before we see a legitimate size get even close...